Add SamplePSDD

RenatoGeh opened the pull request:
Hi,

This PR proposes adding the `SamplePSDD` learning algorithm that is to appear at UAI 2021. If you guys think this is a good fit for Juice, I'd be happy to add the code to the library and make any adjustments to comply with Juice notation or anything like that.

I wrote some tests, but it seems tests were already failing for me before doing anything to the code. I'm not sure if it's because of my GPU (fails at the parameter estimation GPU tests) or if it's something I introduced in the PR. Even though it's in a completely different section of the code, the stack trace says it has something to do with the BDD dependency (which is weird, since it shouldn't have anything to do with it).

The CI also appears to give out an error when running the tests I added. Strangely, the CI log gives no clue whatsoever on what caused them, and running it on my local machine outputs no error on these particular test cases. Running the test files by themselves also gives no errors. I'm surely missing something here, but have no idea what exactly. 

We tried adding a minimal number of dependencies, but ultimately we ended up having to import the following externals:
 - [BinaryDecisionDiagrams.jl](https://github.com/RenatoGeh/BDD.jl): for the BDDs;
 - [ThreadPools.jl](https://github.com/tro3/ThreadPools.jl): for non-uniform threading;
 - [Distributions.jl](https://github.com/JuliaStats/Distributions.jl): for Dirichlet sampling.

To avoid namespace shenanigans, I imported the BDD dependency in its own namespace, since there are quite a few name conflicts with `LogicCircuits`.

As to notation/naming conventions, I somewhat tried mimicking your namings, but not sure if I'm 100% compliant.

The `src/structurelearner/sample_psdd.jl` file implements `SamplePSDD` and any functions it needs. `src/ensembles/ensembles.jl` adds a type `Ensemble` for mixtures of not-necessarily-same-vtree probabilistic circuits. It adds the weight learning methods we mention in the paper (namely EM, likelihood weighting, uniform weights and stacking of mixtures). File `src/ensembles/bmc.jl` adds a `BayesModelComb` type for Bayesian Model Combination of PCs.

Let me know if you want me to change anything in the code.

Thanks

khosravipasha commented:
Thanks for the PR, this should be very useful. I will have some time to review in more details next week or so.

For the tests, the "slow tests" is fine and expected to fail, the "unit tests" portion is currently run on CPU, for the GPU tests we run manually using customized github runner.

Yesh the stack traces are not that informative, and that's why we have a way of running individual tests
https://github.com/Juice-jl/ProbabilisticCircuits.jl/actions/workflows/one_test.yml
You can probably replicated that on your clone and run that action to get better stack trace withing the same CI environment.

RenatoGeh commented:
Thanks for the tips, Pasha. I added a few more tests, added I/O writing for `Ensemble`s and added a function to compile a BDD into a PSDD. That last one's kind of out of place, but since we used that as a baseline for the paper, I added it to the PR as well. Feel free to ignore that commit if you think it's best.

On the issue of tests, I ran the individual tests and they passed with no issues:
- [SamplePSDD](https://github.com/RenatoGeh/ProbabilisticCircuits.jl/runs/2935040870?check_suite_focus=true)
- [Ensembles](https://github.com/RenatoGeh/ProbabilisticCircuits.jl/runs/2935051678?check_suite_focus=true)
- [BMCs](https://github.com/RenatoGeh/ProbabilisticCircuits.jl/runs/2935054664?check_suite_focus=true)
- [BDDs](https://github.com/RenatoGeh/ProbabilisticCircuits.jl/runs/2935043818?check_suite_focus=true)

In the unit tests workflow, these same tests do pass, but they end up [crashing some place else](https://github.com/Juice-jl/ProbabilisticCircuits.jl/pull/79/checks?check_run_id=2935030387). This same (or at least similar) error seems to be occurring on `Juice-jl:master` [as well](https://github.com/Juice-jl/ProbabilisticCircuits.jl/runs/2824432605).

guyvdbroeck commented:
You might want to merge the latest bug fix to master than gets rid of the failing MAP inference test.

RenatoGeh commented:
Great, thanks Guy. Rebased master and now tests seem to pass fine.

khosravipasha reviewed (approved):
Thanks for the PR.
Overall, looks good to me, few comments inside.

1. What's the difference of `BayesModelComb` and `Ensemble`? In terms of storage/types they look the same. We also have SharedProbCircuit (but that's slightly different since the structures are shared). Here I assume each ProbCircuit can have different structure right?
Also might want to rename `Ensemble` slightly, since its for a specific usecase when we have bdd and logical constraints (unless I read it wrong)

2.  I need to double check some interations between the threads and gpu later but if it works now then should be fine. One concern is for bigger circuits having lots of threads might lead to memory issues. Can worry about this later I guess.

It was a lot of code so did not take a full pass yet, but overall looks good to me. Will take another pass early next week, then can merge it.

RenatoGeh commented:
Thanks for the review, Pasha. Answering questions inline.

> Thanks for the PR.
> Overall, looks good to me, few comments inside.
> 
>     1. What's the difference of `BayesModelComb` and `Ensemble`? In terms of storage/types they look the same. 

The difference is that `BayesModelComb` is an ensemble of `Ensemble`s. [Here](https://axon.cs.byu.edu/papers/Kristine.ijcnn2011.pdf)'s the reference paper we based the algorithm on.

> We also have SharedProbCircuit (but that's slightly different since the structures are shared). Here I assume each ProbCircuit can have different structure right?

Yeah. `Ensemble` doesn't assume structure.

>        Also might want to rename `Ensemble` slightly, since its for a specific usecase when we have bdd and logical constraints (unless I read it wrong)

I kept it `Ensemble` since the mixture's weights are learned independent of the components, and so thought it might be useful to keep it as generic as possible. The idea is that we provide a particular case for `SamplePSDD` through `ensemble_sample_psdd`, but we could very well bundle up some SPNs some other way and learn ensemble weights with the existing functions.

> 
>     2. I need to double check some interations between the threads and gpu later but if it works now then should be fine. One concern is for bigger circuits having lots of threads might lead to memory issues. Can worry about this later I guess.
> 
> 
> It was a lot of code so did not take a full pass yet, but overall looks good to me. Will take another pass early next week, then can merge it.

Awesome. :) As a side note, I forgot to add some documentation. I'll write some and push them. I'll mark this WIP until I do.

RenatoGeh commented on src/ensembles/ensembles.jl:
```diff
@@ -0,0 +1,177 @@
+
+"""Split `X` into two partitions `A` and `B`, where `A` is a Bernoulli sample of each element in
+`X` with probability `p` and `B=X∖A`. Guarantees at least one element in `A` and `B`."""
+function bernoulli_partition(X::Vector{Int}, p::Float64)::Tuple{Vector{Int}, Vector{Int}}
```
Alright. Will do.

RenatoGeh commented on src/ensembles/ensembles.jl:
```diff
@@ -0,0 +1,177 @@
+    2. `:uniform` for uniform weights;
+    3. `:em` for Expectation-Maximization;
+    4. `:stacking` for mixture model Stacking;
+"""
```
UAI is still on preliminaries. But I'll add them once UAI releases the final version. Thanks.

RenatoGeh commented on src/structurelearner/bdd.jl:
```diff
@@ -0,0 +1,76 @@
+    return passdown(U)
+end
+
+"Returns a structured probabilistic circuit compiled from a binary decision diagram."
```
Should work with any BDD. I guess a restriction is that it's limited to the BDD.jl implementation: we assume a fixed (arbitrary) ordering, and so the PSDD will compile from a particularly ordered BDD.

RenatoGeh commented on src/ensembles/ensembles.jl:
```diff
@@ -0,0 +1,177 @@
+    return F
+end
+
+"Learns the weights of the Ensemble by Stacking, with `k` as the number of folds in k-fold."
```
Yeah, that's the idea, but in this case we used a density estimation version. [Here](https://papers.nips.cc/paper/1997/file/ee8374ec4e4ad797d42350c904d73077-Paper.pdf)'s the reference paper.

guyvdbroeck commented:
Would it be realistic to move the BinaryDecisionDiagrams.jl dependency and related functionality to LogicCircuits.jl? It already has an SDD compiler (which of course can also compile BDDs by fixing a right-linear vtree). I would prefer to keep all purely logical features there.

RenatoGeh commented:
Do you mean moving the BDD code into LogicCircuits.jl? Or re-implementing the code using LogicCircuits.jl syntax through the existing types? I see some problems with the latter, since SamplePSDD relies on efficiently reducing, restricting and applying BDDs, which could introduce some overhead if we try to generalize with logic circuits. In the case of the former, I'd personally see no problem in moving BinaryDecisionDiagrams.jl to LogicCircuits.jl. Although one issue would be that the code would be somewhat disconnected from the rest.

guyvdbroeck commented:
I would prefer to move the BDD code into LogicCircuits.jl. Perhaps in future then we can have some interoperability between your BDD code and the other logic circuit types, some conversions, support for the LogicCircuits.jl compilation API, etc.

RenatoGeh commented:
Yeah, I see no problem with that. I'll send the PR to LogicCircuits.jl.

khosravipasha commented:
@RenatoGeh 
Recently upgraded the DataFrames dependency to 1.x. In this new version, there is a breaking change that cannot do `DataFrame(matrix)` anymore, and need to replace it with `DataFrame(matrix, :auto)`. In case you ran into any issues.

Also the LogicCircuit PR is now pushed and in master.

RenatoGeh commented:
> @RenatoGeh
> Recently upgraded the DataFrames dependency to 1.x. In this new version, there is a breaking change that cannot do `DataFrame(matrix)` anymore, and need to replace it with `DataFrame(matrix, :auto)`. In case you ran into any issues.
> 
> Also the LogicCircuit PR is now pushed and in master.

Thanks for heads up, Pasha! I did have one issue when writing the docs. I wanted to show how to build a PSDD from a CNF file and ran into #80. Would you rather we deal with this in another PR and try to merge this right now, or try to resolve this in this one?

khosravipasha commented:
> Thanks for heads up, Pasha! I did have one issue when writing the docs. I wanted to show how to build a PSDD from a CNF file and ran into #80. Would you rather we deal with this in another PR and try to merge this right now, or try to resolve this in this one?

@RenatoGeh Yes this seems to be a separate issue, we can merge this one first and do the fix and extra docs in future commits. Thanks.

RenatoGeh commented:
Ok, thanks. Refactored everything to use `LogicCircuits`' BDDs now and added a very simple how-to in the manuals. From what I gather, tests are not passing because `LogicCircuits`' version set in `Project.toml` points to version `0.2.3`, which doesn't have the recent required changes.

khosravipasha commented:
Thanks Renato, will take a look at the tests, probably config issue cause
the tests should also grab logic circuits from master in dev mode. I will
just run them locally if not. We have been meaning to release a new version
anyway.

khosravipasha commented:
Fixed the tests errors, ran it locally and it passed, will let tests run here too, and will merge sometime next week if there is no other changes needed.

RenatoGeh commented:
Alright, awesome. Thanks for that last fix, review and feedback. Just one last thing about the docs. I mention how to compile a PSDD from an SDD, but as we know that's broken for now. Just giving you a heads up in case we have anyone asking about it in the issues.

khosravipasha commented:
Great, thanks. Yeah for the SDD => PSDD, is it that same as the issue https://github.com/Juice-jl/ProbabilisticCircuits.jl/issues/80? 

Will merge this PR now, if there is anything else we can make extra issues.

RenatoGeh commented:
> Great, thanks. Yeah for the SDD => PSDD, is it that same as the issue #80?

It is.
> 
> Will merge this PR now, if there is anything else we can make extra issues.

Sure.
